extensions/ql-vscode/src/queries-panel/query-pack-discovery.ts

```diff
@@ -0,0 +1,105 @@
+import { basename, dirname } from "path";
+import { CodeQLCliServer, QuerySetup } from "../codeql-cli/cli";
+import { Event } from "vscode";
+import { dbSchemeToLanguage } from "../common/query-language";
+import { FALLBACK_QLPACK_FILENAME, QLPACK_FILENAMES } from "../pure/ql";
+import { FilePathDiscovery } from "../common/vscode/file-path-discovery";
+import { getErrorMessage } from "../pure/helpers-pure";
+import { extLogger } from "../common";
+import { EOL } from "os";
+import { containsPath } from "../pure/files";
+import { getOnDiskWorkspaceFolders } from "../common/vscode/workspace-folders";
+
+export interface QueryPack {
+  path: string;
+  language: string | undefined;
+}
+
+/**
+ * Discovers all query packs in the workspace.
+ */
+export class QueryPackDiscovery extends FilePathDiscovery<QueryPack> {
+  constructor(private readonly cliServer: CodeQLCliServer) {
+    super("Query Pack Discovery", `**/{${QLPACK_FILENAMES.join(",")}}`);
+  }
+
+  /**
+   * Event that fires when the set of query packs in the workspace changes.
+   */
+  public get onDidChangeQueryPacks(): Event<void> {
+    return this.onDidChangePathsEmitter.event;
+  }
+
+  /**
+   * Given a path of a query file, locate the query pack that contains it and
+   * return the language of that pack. Returns undefined if no pack is found
+   * or the pack's language is unknown.
+   */
+  public getLanguageForQueryFile(queryPath: string): string | undefined {
+    // Find all packs in a higher directory than the query
+    const packs = this.paths.filter((queryPack) =>
+      containsPath(dirname(queryPack.path), queryPath),
+    );
+
+    // Sort by descreasing path length to find the pack nearest the query
+    packs.sort((a, b) => b.path.length - a.path.length);
+
+    if (packs.length === 0) {
+      return undefined;
+    }
+
+    // If the first two packs are from the same directory then look at the filenames
+    if (
+      packs.length >= 2 &&
+      dirname(packs[0].path) === dirname(packs[1].path)
+    ) {
```

Comment:
Can we make this easier to read by splitting this into two if-conditions and returning early? Something like:

```ts
    if (packs.length === 1) {
      return packs[0].language;
    }

    // If the first two packs are from a different directory, then the first one is the nearest
    if (dirname(packs[0].path) !== dirname(packs[1].path)) {
      return packs[0].language;
    }

    // If the first two packs are from the same directory then look at the filenames
    if (basename(packs[0].path) === FALLBACK_QLPACK_FILENAME) {
      return packs[0].language;
    }

    return packs[1].language;
```